src/coreclr/vm/gcenv.ee.cpp

```diff
@@ -444,6 +444,36 @@ gc_alloc_context * GCToEEInterface::GetAllocContext()
     return pThread->GetAllocContext();
 }
 
+void InvokeGCAllocCallback(ee_alloc_context* pEEAllocContext, enum_alloc_context_func* fn, void* param)
+{
+
+    gc_alloc_context* pAllocContext = &pEEAllocContext->gc_alloc_context;
+
+    // The allocation context might be modified by the callback, so we need to save
+    // the remaining sampling budget and restore it after the callback.
+    size_t remainingSamplingBudget = (pEEAllocContext->alloc_sampling == nullptr) ? 0 : pEEAllocContext->alloc_sampling - pAllocContext->alloc_ptr;
+
+    fn(pAllocContext, param);
+
+    if (remainingSamplingBudget != 0)
+    {
+        // fast_alloc_limit_ptr should be in the range [alloc_ptr, alloc_limit].
+        if (remainingSamplingBudget > (size_t)(pAllocContext->alloc_limit - pAllocContext->alloc_ptr))
```

Comment:
I've thought about that and I'm not sure which way is really skewing the statistics...
The odds to be in an AC are lower than not being in an AC so if we don't count (because we don't send the event), it should impact more the distribution than recomputing (and so don't count the previous sample).
Also, I don't know what the GC will do with the ACs: will the size of the AC change? If not, then this code is fine because it is just shifting the threshold if ptr changed.

Reply:
This is a little math proof the behavior I'm worried about. If you think there is an error in the argument of course let me know but so far I don't see one. For the sake of the proof I'm assuming the code inside the if statement would have recomputed the limit as described in the comment as one of the options. The other option of setting `alloc_sampling=alloc_ptr` is also problematic though the exact way in which it skews the statistical distribution is different.

 
In order to give the correct statistics we must maintain an invariant that there is a `(1-p)^N * p` chance for `alloc_sampling` to have value `alloc_ptr+N` for all `N` within the AC range. Assume that when we entered this function the invariant was true (otherwise we've got bugs elsewhere). If the invariant is true for all `N`, it must be true for `N=1`, therefore we can safely assume that on function entrance `alloc_sampling==alloc_ptr+1` with probability `(1-p)*p`.  Now we can consider what is the probability that `alloc_sampling==alloc_ptr+1` at the end of the function? If it isn't also `(1-p)*p` then we've broken the invariant we needed for correct statistics.

 Let `new_AC_size = pAllocContext->alloc_limit - pAllocContext->alloc_ptr` after GC callback was invoked. There is some undetermined but non-zero probability X that `remainingSamplingBudget > new_AC_size > 1`, otherwise it would be pointless to have an if() in the code that checks for a case that never occurs. So there are two ways we can end up with `alloc_sampling==alloc_ptr+1`:

1) If `alloc_sampling==alloc_ptr+1` on function entry then remainingSamplingBudget = 1, we go to the else branch and set `alloc_sampling==alloc_ptr+1`
2) If `remainingSamplingBudget > new_AC_size > 1` then we could recompute a new threshold randomly and our random number happens to pick `alloc_sampling=alloc_ptr+1`

The odds of case (1) happening is `(1-p)*p`. The odds of case (2) happening is `X*(1-p)*p`. The cases are mutually exclusive so the chance that either of them happens is the sum `(1-p)*p + X*(1-p)*p`. We know `X!=0` which means this probability is not the same as our desired probability `(1-p)*p`.

Does that help show why this doesn't work?

Reply:
We should sum the probabilities of the 2 branches for the odds of having `alloc_sampling==alloc_ptr+1` at the end of the function to check the invariant:
branch 1 =` (1-X)(1-p)*p`
+
branch 2 = `X(1-p)*p`
= `(1-X + X)(1-p)*p`
= `(1-p)*p`
So whatever we will be doing, the invariant stays invariant.

My concern is different: we used the exponential formula to compute the next byte to sample and the probability to end up into an 8KB allocation context is the integral from x=0 to 8192 of 1/100K * exp(-1/100K * x) ~= 0.078 ; less than 8% so if we reroll the dice and forget the previous sampled byte, it has more impact on the distribution (= less samples) than keeping alloc_sampling inside the current allocation context.

If you forget about `alloc_sampling` as a pointer for a moment, and replace it by a boolean saying "we need to sample that allocation context", what would we want to do?
Since your comment, I've read some mathematical books and discussed with colleagues who implemented the sampling in our products. I think we don't count the same things. Using an exponential distribution to compute the distance between 2 sampled bytes is equivalent to the Head/Tail counting: I hope we agree on that  :^)
Our difference is that changing the probability of the next byte to sample (in or out of an allocation context) does not have the 1/2 probability of H/T and we can't use the same p * (1-p)^N type of computation because it should be seen as p1 * p2 * ... * pN+1 with px =  1/100K * exp(-1/100K * x). And again, we are not talking about the same probability!

As a conclusion, no problem to change the code and recompute the `alloc_sampling` if the AC was sampled and its new size is smaller than `alloc_sampling`: this should not happen that much and so don't substancially change the distribution of sampled bytes.

Reply:
> As a conclusion, no problem to change the code and recompute the alloc_sampling if the AC was sampled and its new size is smaller than alloc_sampling: this should not happen that much and so don't substancially change the distribution of sampled bytes.

I think we still disagree unless you are talking about some recomputation that is different than the ones we've been talking about previously. We have options to implement this that (to the best of my current understanding) do not introduce any bias into the sampling. I do not want to pick an implementation that biases the results a small amount when it would have been just as easy to implement it without biasing the result at all. These are the implementations I believe do not bias the result:
- Recompute always (regardless of the prior value of alloc_sampling)
- Recompute if the AC changed size (regardless of the prior value of alloc_sampling)
- If the AC previously contained the sampling threshold and the GC shrinks it below the threshold size, treat it as an unsampled AC.

Conditionally recomputing it only if the AC was previously sampled and has now shrunk does bias the results and for now at least I still stand behind the reasoning I gave above. I read through what you wrote and more detailed responses to that is below...


> branch 1 = (1-X)(1-p)*p

I believe this is incorrect. On entrance to this function there was a probability `(1-p)*p` chance that `alloc_sampling==alloc_ptr+1`. I see no code path where `alloc_sampling` has that value initially and doesn't go into branch 1. This means the probability of hitting branch 1 is also `(1-p)*p`. 

(My argument did only look at the case where the GC returns an AC size > 1 though I didn't explicitly mention it. If you believe the handling of AC sizes <= 1 is where the error is I could elaborate on it, but in my head that case also checks out.)

> if we reroll the dice and forget the previous sampled byte, it has more impact on the distribution (= less samples) than keeping alloc_sampling inside the current allocation context.

I don't follow your argument? I agree that regenerating a new random number would give approximately 7.8% probability for a 8192 byte AC to be sampled. I don't see how that fact supports your conclusion that 'it has more impact on the distribution'.

I think you might be focusing on the possibility that before the GC the AC may have had the sampling threshold inside of it and after regenerating a new random threshold now it doesn't. That is definitely possible and I'd expect it to occur with probability:
```
Pr(pre-GC AC would have been sampled) * Pr(post-GC recomputed AC is not sampled)
0.078                                 * (1-0.078)
0.0719
```
However if we are recomputing always that means the opposite pattern will also occur - the pre-GC AC wasn't being sampled and after recomputing a new random number now it is. This occurs with the exact same probability and two cases cancel each other out:
```
Pr(pre-GC AC would not have been sampled) * Pr(post-GC recomputed AC is sampled)
(1-0.078)                                 * 0.078
0.0719
```

> Using an exponential distribution to compute the distance between 2 sampled bytes is equivalent to the Head/Tail counting: I hope we agree on that :^)

Agreed... with a pedantic note that I assume you are ignoring for the moment the minor differences between Geometric and Exponential distributions :)

> Our difference is that changing the probability of the next byte to sample (in or out of an allocation context) does not have the 1/2 probability of H/T 

I'm not sure what you meant here. I agree we aren't planning to set p=1/2 and whatever we do needs to work for all values of p. However I'm not aware that anything in my arguments was based on p being 1/2. For all examples involving coins I assume we are flipping weighted coins that probability p of heads. If you believe something in my argument relied on p=1/2 it would help to point out one specific step in the argument where that dependency occurs.

I think when you say "changing the probability of the next byte to sample" you mean generating a new random threshold but I'm not certain? If we are doing this correctly we should never be changing the probability that the next byte gets sampled, that probability should always be `p`.

> and we can't use the same p * (1-p)^N type of computation because it should be seen as p1 * p2 * ... * pN+1 with px = 1/100K * exp(-1/100K * x)

I'm not sure what you meant here? I think you defined `px` as the PDF of the exponential distribution but then evaluated it on a bunch of integers and multiplied those results together when calculating p1*p2*...*pN+1? It doesn't seem to be calculating the probability of anything we'd need to care about? Perhaps you were intending to sum or to integrate that sequence instead?